Paco-Server/src/com/google/sampling/experiential/server/InternalRequestIdentificationFilter.java

```diff
@@ -0,0 +1,58 @@
+package com.google.sampling.experiential.server;
+
+import java.io.IOException;
+import java.util.logging.Logger;
+
+import javax.servlet.Filter;
+import javax.servlet.FilterChain;
+import javax.servlet.FilterConfig;
+import javax.servlet.ServletException;
+import javax.servlet.ServletRequest;
+import javax.servlet.ServletResponse;
+import javax.servlet.http.HttpServletRequest;
+
+import com.google.apphosting.api.ApiProxy;
+
+public final class InternalRequestIdentificationFilter implements Filter {
+  public static final Logger log = Logger.getLogger(InternalRequestIdentificationFilter.class.getName());
+  //TODO: Should we respond with different messages for each of the scenario ??
+  private static final String INVALID_HEADER = "Unauthorized : Invalid Header";
+  private static final String INVALID_HEADER_MISMATCH = "Unauthorized : Header not matching";
+  private static final String INVALID_HOST = "Unauthorized : Invalid host name";
+  
+  @Override
+  public void doFilter(ServletRequest req, ServletResponse resp, FilterChain chain) throws IOException,
+                                                                                   ServletException {
+    String appIdFromHeader = ((HttpServletRequest)req).getHeader("X-Appengine-Inbound-Appid");
+    log.info("IRI filter");
+    ApiProxy.Environment env = ApiProxy.getCurrentEnvironment();
+    
+    if (appIdFromHeader == null) {
+      log.warning(INVALID_HEADER);
+      resp.getWriter().write(INVALID_HEADER);
+      return;
+    }
+    String appIdFullHostName = env.getAttributes().get("com.google.appengine.runtime.default_version_hostname").toString();
+   
+    if (appIdFullHostName != null) {
+      String[] appIdPartialHostName = appIdFullHostName.split("\\.");
+      String appIdFromRuntime = appIdPartialHostName[0];
+      if (!appIdFromHeader.equalsIgnoreCase(appIdFromRuntime)) {
+        log.warning(INVALID_HEADER_MISMATCH);
+        resp.getWriter().write(INVALID_HEADER_MISMATCH);
+      }
+    } else {
+      log.warning(INVALID_HOST);
+      resp.getWriter().write(INVALID_HOST);
+    }
+    chain.doFilter(req, resp);
```

Comment:
Is it enough just to check if the header exists, or, do we need to check that it has a value that is unique to our cloud project?

Reply:
We are not using this filter anymore. Since the admin role seems to work with the upgrade in staging environments.

Reply:
Making sure I understand. The admin security constraint on backend modules allows other modules within our service to connect and allows users who are admins in the service to connect directly but prevents anyone else and any other modules in the Google cloud (outside of our service) from connecting? 

If all that is true, then there is only one other bit to decide: should we use this X-Appengine header to make sure that only modules can call the endpoint (meaning it would prevent direct browser requests to it)? I can see the utility of an admin test hitting an endpoint but I can also see that it confuses the issue since it is only supposed to be accessible by another module in practice.

Reply:
All that is true. Also, The header does not get set in the local environment. So, if we want to have this check, between staging and dev we need to support two different versions of web.xml.

Reply:
You convinced me with that problem that we should leave it as is :-)